### ISSUE TITLE
Don't call callbacks from this.setState() to avoid double rendering

### ISSUE THREAD

**Issue by eldargab** [wojtekmaj/react-calendar#298]
Hello, I noticed that react-calendar notifies parent component about various events via `.setState()` callback. 

E.g.

https://github.com/wojtekmaj/react-calendar/blob/be7c443fbe39ce90036380dab727c1ca7072a073/src/Calendar.jsx#L310

The problem is that`.setState()` callback is called after react finishes real DOM updates. That means, changes from parent component will cause another update cycle. Also parent component might remove calendar in response to event entirely making previous update completely useless.

You can verify such behaviour via DevTools profiler

<img width="1137" alt="image" src="https://user-images.githubusercontent.com/762157/71369290-5564c180-25bb-11ea-96d3-302e5d313e64.png">

In the screenshot above all the work before `Lifecycle hook scheduled a cascading update` is completely useless and causes substantial delay.

Another reason not to use `.setState()` for parent callbacks is to avoid crashing entire app in case  parent event handler erroneously raises exception.

**Comment by wojtekmaj** [comment]
Hmmm, if we moved this to the beginning of the function then we would potentially start two asynchronous updates - one via onChange & props, one internal via state. That would be even worse, I think?

**Comment by eldargab** [comment]
No, there would be only one update incorporating both props and state changes.

This is what currently happens with [react-calendar/sample/parcel](/wojtekmaj/react-calendar/tree/master/sample/parcel) app.

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/762157/71658932-1bd04b00-2d5f-11ea-937e-e0d089172c29.png">

On other hand, immediate invocation of `onChange` callback yields the following picture

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/762157/71659203-293a0500-2d60-11ea-9e66-6cb5ad79f578.png">

PR #302 illustrates proposed changes.

**Comment by wojtekmaj** [comment]
This no longer applies since Calendar was rewritten to hooks and the old `setState(..., callback)` path is gone. Closing.
